### PR TITLE
issue: 825055 "Dummy Send" for improving low msg rate latency

### PR DESCRIPTION
**Pull request by liranoz12**
This feature is to add new API to allow application to call a dummy send
flow which will warm up the CPU caches on the send fast path.
Enable using DUMMY_WARM_MSG in the flags field of send(), sendto(), sendmsg().

Signed-off-by: Liran Oz <lirano@mellanox.com>

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/659/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/651/ for details (Mellanox internal link).

File: src/stats/stats_printer.cpp [hunk 1]
```diff
@@ -230,6 +230,14 @@ void print_full_stats(socket_stats_t* p_si_stats, mc_grp_info_t* p_mc_grp_info,
 		fprintf(filename, "Tx OS info: %u KB / %u / %u / %u [bytes/packets/eagains/errors]%s\n",  p_si_stats->counters.n_tx_os_bytes/BYTES_TRAFFIC_UNIT,  p_si_stats->counters.n_tx_os_packets, p_si_stats->counters.n_tx_os_eagain, p_si_stats->counters.n_tx_os_errors, post_fix);
 		b_any_activiy = true;
 	}
+	if (p_si_stats->counters.n_tx_dummy) {
+		fprintf(filename, "Tx Dummy messages : %d\n", p_si_stats->counters.n_tx_dummy);
+		b_any_activiy = true;
+	}
+	if (p_si_stats->counters.n_tx_dummy_drops) {
```
Comment (OphirMunk) [inline src/stats/stats_printer.cpp]: rename drop to abort

Comment (OphirMunk) [inline src/stats/stats_printer.cpp]: remove drop counter and print

File: src/vma/dev/qp_mgr.h [hunk 2]
```diff
@@ -147,6 +151,8 @@ class qp_mgr
 	uint32_t 		m_rx_num_wr;
 	uint32_t 		m_tx_num_wr;
 
+	bool  			m_hw_dummy_send_support;
```
Comment (OphirMunk) [inline src/vma/dev/qp_mgr.h]: Assuming ibv_dev_ctx... will check dummy capability and passes it to ring then m_hw_dummy_send_support can be const and the check will only be in device creation and not per ring creation

File: src/vma/dev/ring_bond.cpp [hunk 3]
```diff
@@ -239,6 +239,22 @@ void ring_bond::send_lwip_buffer(ring_user_id_t id, vma_ibv_send_wr* p_send_wqe,
 	}
 }
 
+bool ring_bond::get_hw_dummy_send_support(ring_user_id_t id, vma_ibv_send_wr* p_send_wqe) {
```
Comment (OphirMunk) [inline src/vma/dev/ring_bond.cpp]: Open bracket - move to next line

File: src/vma/dev/ring_bond.cpp [hunk 4]
```diff
@@ -239,6 +239,22 @@ void ring_bond::send_lwip_buffer(ring_user_id_t id, vma_ibv_send_wr* p_send_wqe,
...
+	} else {
+		active_ring = m_bond_rings[id];
+		if (likely(p_mem_buf_desc->p_desc_owner == active_ring)) {
+			return active_ring->get_hw_dummy_send_support(id, p_send_wqe);
+		} else {
+			return false;
+		}
+	}
+}
+
```
Comment (OphirMunk) [inline src/vma/dev/ring_bond.cpp]: In next implementation where will send dummy anyway - this function should be removed (failure will be detected at the very low level)

File: src/vma/dev/wqe_send_handler.h [hunk 5]
```diff
@@ -47,6 +47,16 @@ class wqe_send_handler: public tostr
 	void init_wqe(vma_ibv_send_wr &wqe_to_init, struct ibv_sge* sge_list, uint32_t num_sge);
 	void init_inline_wqe(vma_ibv_send_wr &wqe_to_init, struct ibv_sge* sge_list, uint32_t num_sge);
 
+	inline vma_ibv_wr_opcode set_dummy_opcode(vma_ibv_send_wr &wqe) {
```
Comment (OphirMunk) [inline src/vma/dev/wqe_send_handler.h]: rename set_dummy_opcde dummy_opcode_set/unset

File: src/vma/dev/wqe_send_handler.h [hunk 6]
```diff
@@ -47,6 +47,16 @@ class wqe_send_handler: public tostr
...
+		vma_ibv_wr_opcode last_opcode = vma_send_wr_opcode(wqe);
+		vma_send_wr_opcode(wqe) = VMA_IBV_WR_NOP;
+		return last_opcode;
+	}
+
+	inline void unset_dummy_opcode(vma_ibv_send_wr &wqe, vma_ibv_wr_opcode opcode) {
```
Comment (OphirMunk) [inline src/vma/dev/wqe_send_handler.h]: Single function implementation: set_wr_opcode (which returns previous opcode)

**Comment by liranoz12** [comment]
@rosenbaumalex @OphirMunk  Please review

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/670/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/662/ for details (Mellanox internal link).

**Comment by OphirMunk** [comment]
Merged #254

**Pull request closed without merging**